hazelcast/src/main/java/com/hazelcast/sql/impl/expression/ConstantExpression.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+import com.hazelcast.sql.impl.type.converter.Converter;
+import com.hazelcast.sql.impl.type.converter.Converters;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Constant expression.
+ *
+ * @param <T> Return type.
+ */
+public final class ConstantExpression<T> implements Expression<T>, IdentifiedDataSerializable {
+
+    private QueryDataType type;
+    private T value;
+
+    public ConstantExpression() {
+        // No-op.
+    }
+
+    private ConstantExpression(T value, QueryDataType type) {
+        this.type = type;
+        this.value = value;
+    }
+
+    public static ConstantExpression<?> create(Object value, QueryDataType type) {
+        if (type.getTypeFamily() == QueryDataTypeFamily.NULL) {
+            assert value == null;
+            return new ConstantExpression<>(null, QueryDataType.NULL);
+        }
+
+        if (value == null) {
+            return new ConstantExpression<>(null, type);
+        }
+
+        Converter valueConverter = Converters.getConverter(value.getClass());
+        Converter typeConverter = type.getConverter();
+        value = typeConverter.convertToSelf(valueConverter, value);
```

Comment:
Do we have tests that simulate an exception that happened during this conversion? What would the user receive in this case?

Reply:
This conversion logic should never throw under normal conditions. The type of a literal and possibility of the conversion is already verified during query validation, see `HazelcastSqlValidator.validateLiteral` and `HazelcastTypeSystem#canRepresent`. In other words, we already attempted to perform the conversion during validation and if we have entered this method that means the conversion is valid, otherwise the query validation fails and we are obviously not creating any expressions for the query. If this conversion logic throws, that means the validation is broken, but that's kind of a system error, not the error user should deal with normally.